multiphenicsx/fem/petsc.py

```diff
@@ -1573,34 +1574,67 @@ def set_bc(  # type: ignore[no-any-unimported]
     b: petsc4py.PETSc.Vec, bcs: list[dolfinx.fem.DirichletBC] = [],
     x0: typing.Optional[petsc4py.PETSc.Vec] = None,
     alpha: float = 1.0,
-    restriction: typing.Optional[mcpp.fem.DofMapRestriction] = None
+    restriction: typing.Optional[mcpp.fem.DofMapRestriction] = None,
+    restriction_x0: typing.Optional[mcpp.fem.DofMapRestriction] = None,
 ) -> None:
-    """Apply the function :func:`dolfinx.fem.set_bc` to a PETSc Vector."""
+    """Apply the function :func:`dolfinx.fem.set_bc` to a PETSc Vector:
+
+    For each Dirichet condition in bcs, it modifies b such that:
+
+    .. math::
+
+        b \\leftarrow  \\alpha (u_{bc} - x_0)
+
+    where :math:`u_{bc}` contains the Dirichlet values.
+
+    Parameters
+    ----------
+        b
+            PETSc Vector
+        bcs
+            Dirichlet conditions
+        x0
+            PETSc Vector subtracted to the Dirichlet values.
+        alpha
+            Scaling factor
+        restriction, restriction_x0 
+            Dofmap restrictions for b and x0. If not provided, the unrestricted vectors will be used for each.
+    """
+    @contextmanager
+    def x0_wrapper():
+        try:
+            if x0 is None:
+                yield x0
+            elif restriction_x0 is None:
+                yield x0.array_r
+            else:
+                with VecSubVectorReadWrapper(x0, restriction_x0.dofmap, restriction_x0, ghosted=False) as x0_sub:
+                    yield x0_sub
+        finally:
+            pass
     if restriction is None:
-        if x0 is not None:
-            x0 = x0.array_r
-        for bc in bcs:
-            bc.set(b.array_w, x0, alpha)
+        with x0_wrapper() as x0_view:
+            for bc in bcs:
+                bc.set(b.array_w, x0_view, alpha)
     else:
         with VecSubVectorWrapper(b, restriction.dofmap, restriction, ghosted=False) as b_sub, \
-                VecSubVectorReadWrapper(x0, restriction.dofmap, restriction, ghosted=False) as x0_sub:
-            for bc in bcs:
-                bc.set(b_sub, x0_sub, alpha)
-
+                x0_wrapper() as x0_view:
+                    for bc in bcs:
+                        bc.set(b_sub, x0_view, alpha)
 
 def set_bc_nest(  # type: ignore[no-any-unimported]
     b: petsc4py.PETSc.Vec, bcs: list[list[dolfinx.fem.DirichletBC]] = [],
     x0: typing.Optional[petsc4py.PETSc.Vec] = None,
     alpha: float = 1.0,
-    restriction: typing.Optional[list[mcpp.fem.DofMapRestriction]] = None
+    restriction: typing.Optional[list[mcpp.fem.DofMapRestriction]] = None,
+    restriction_x0: typing.Optional[list[mcpp.fem.DofMapRestriction]] = None
 ) -> None:
     """Apply the function :func:`dolfinx.fem.set_bc` to each sub-vector of a nested PETSc Vector."""
     if restriction is None:
         dofmaps = [None] * len(b.getNestSubVecs())
     else:
         dofmaps = [restriction_.dofmap for restriction_ in restriction]
     dofmaps_x0 = dofmaps
-    restriction_x0 = restriction
```

Comment:
I am happy with this, I am just wondering if there is any case in which they would really be different. I guess one would need `restriction_x0 != restriction` when assembling a rectangular matrix, right? (although I have never had anyone do that, at least to my knowledge).

Reply:
I am not working with multiple restrictions on a single domain. Hard to imagine where this would be used but on the bright side it is a good sign that the code is this flexible.

The only scenario I can imagine is if you had the solution on a subdomain and wanted to solve a problem on a subsubdomain.